<a href="https://colab.research.google.com/github/HamburgermanA3445/heartdiseaseprediction/blob/main/HeartDisease_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Unsuccessful, obtained from Kaggle:

https://www.kaggle.com/datasets/kamilpytlak/personal-key-indicators-of-heart-disease

*some of the cells at the end were not run

In [ ]:
!unzip "/content/kamil_pytlak (4).zip"

Archive:  /content/kamil_pytlak (4).zip
  inflating: heart_disease_health_indicators_BRFSS2015.csv  


In [ ]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder

In [ ]:
heart_data = pd.read_csv("/content/heart_disease_health_indicators_BRFSS2015.csv")

In [ ]:
label_encoders = {}

for col in heart_data.select_dtypes(include=['object']).columns:
    le = LabelEncoder()
    heart_data[col] = le.fit_transform(heart_data[col])
    label_encoders[col] = le

In [ ]:
heart_data.head()

HeartDiseaseorAttack  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                   0.0     1.0       1.0        1.0  40.0     1.0     0.0   
1                   0.0     0.0       0.0        0.0  25.0     1.0     0.0   
2                   0.0     1.0       1.0        1.0  28.0     0.0     0.0   
3                   0.0     1.0       0.0        1.0  27.0     0.0     0.0   
4                   0.0     1.0       1.0        1.0  24.0     0.0     0.0   

   Diabetes  PhysActivity  Fruits  ...  AnyHealthcare  NoDocbcCost  GenHlth  \
0       0.0           0.0     0.0  ...            1.0          0.0      5.0   
1       0.0           1.0     0.0  ...            0.0          1.0      3.0   
2       0.0           0.0     1.0  ...            1.0          1.0      5.0   
3       0.0           1.0     1.0  ...            1.0          0.0      2.0   
4       0.0           1.0     1.0  ...            1.0          0.0      2.0   

   MentHlth  PhysHlth  DiffWalk  Sex   Age  Education  Income  
0      18.0      15.0       1.0  0.0   9.0        4.0     3.0  
1       0.0       0.0       0.0  0.0   7.0        6.0     1.0  
2      30.0      30.0       1.0  0.0   9.0        4.0     8.0  
3       0.0       0.0       0.0  0.0  11.0        3.0     6.0  
4       3.0       0.0       0.0  0.0  11.0        5.0     4.0  

[5 rows x 22 columns]

In [ ]:
target = 'HeartDiseaseorAttack'
all_columns = heart_data.columns.drop(target)

In [ ]:
heart_data.columns

Index(['HeartDiseaseorAttack', 'HighBP', 'HighChol', 'CholCheck', 'BMI',
       'Smoker', 'Stroke', 'Diabetes', 'PhysActivity', 'Fruits', 'Veggies',
       'HvyAlcoholConsump', 'AnyHealthcare', 'NoDocbcCost', 'GenHlth',
       'MentHlth', 'PhysHlth', 'DiffWalk', 'Sex', 'Age', 'Education',
       'Income'],
      dtype='object')

In [ ]:
heart_data.head()

HeartDiseaseorAttack  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                   0.0     1.0       1.0        1.0  40.0     1.0     0.0   
1                   0.0     0.0       0.0        0.0  25.0     1.0     0.0   
2                   0.0     1.0       1.0        1.0  28.0     0.0     0.0   
3                   0.0     1.0       0.0        1.0  27.0     0.0     0.0   
4                   0.0     1.0       1.0        1.0  24.0     0.0     0.0   

   Diabetes  PhysActivity  Fruits  ...  AnyHealthcare  NoDocbcCost  GenHlth  \
0       0.0           0.0     0.0  ...            1.0          0.0      5.0   
1       0.0           1.0     0.0  ...            0.0          1.0      3.0   
2       0.0           0.0     1.0  ...            1.0          1.0      5.0   
3       0.0           1.0     1.0  ...            1.0          0.0      2.0   
4       0.0           1.0     1.0  ...            1.0          0.0      2.0   

   MentHlth  PhysHlth  DiffWalk  Sex   Age  Education  Income  
0      18.0      15.0       1.0  0.0   9.0        4.0     3.0  
1       0.0       0.0       0.0  0.0   7.0        6.0     1.0  
2      30.0      30.0       1.0  0.0   9.0        4.0     8.0  
3       0.0       0.0       0.0  0.0  11.0        3.0     6.0  
4       3.0       0.0       0.0  0.0  11.0        5.0     4.0  

[5 rows x 22 columns]

#Logistic Model

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [ ]:
#Use for filtering categorical data into numerical
for l in all_columns:
  all_x = []
  for x in heart_data[l]:
    if x not in all_x:
      all_x.append(x)
  print(f'{l}: {all_x}')

age: [40, 49, 37, 48, 54, 39, 45, 58, 42, 38, 43, 60, 36, 44, 53, 52, 51, 56, 41, 32, 65, 35, 59, 50, 47, 31, 46, 57, 55, 63, 66, 34, 33, 61, 29, 62, 28, 30, 74, 68, 72, 64, 69, 67, 73, 70, 77, 75, 76, 71]
sex: [1, 0]
chest pain type: [2, 3, 4, 1]
resting bp s: [140, 160, 130, 138, 150, 120, 110, 136, 115, 100, 124, 113, 125, 145, 112, 132, 118, 170, 142, 190, 135, 180, 108, 155, 128, 106, 92, 200, 122, 98, 105, 133, 95, 80, 137, 185, 165, 126, 152, 116, 0, 144, 154, 134, 104, 139, 131, 141, 178, 146, 158, 123, 102, 96, 143, 172, 156, 114, 127, 101, 174, 94, 148, 117, 192, 129, 164]
cholesterol: [289, 180, 283, 214, 195, 339, 237, 208, 207, 284, 211, 164, 204, 234, 273, 196, 201, 248, 267, 223, 184, 288, 215, 209, 260, 468, 188, 518, 167, 224, 172, 186, 254, 306, 250, 177, 227, 230, 294, 264, 259, 175, 318, 216, 340, 233, 205, 245, 194, 270, 213, 365, 342, 253, 277, 202, 297, 225, 246, 412, 265, 182, 218, 268, 163, 529, 100, 206, 238, 139, 263, 291, 229, 307, 210, 329, 147, 85, 269, 27

In [ ]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score, f1_score

In [ ]:
X = heart_data[all_columns]
y = heart_data[target]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
lg = LogisticRegression()
lg.fit(X_train, y_train)
y_pred = lg.predict(X_test)
accuracy_score(y_test, y_pred)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.9065752128666036

In [ ]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.92      0.99      0.95     45968
         1.0       0.51      0.13      0.21      4768

    accuracy                           0.91     50736
   macro avg       0.71      0.56      0.58     50736
weighted avg       0.88      0.91      0.88     50736



In [ ]:
coefficients = lg.coef_[0]
feature_importance = np.abs(coefficients)

importance = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': coefficients,
    'Importance': feature_importance
})

importance = importance.sort_values(by='Importance', ascending=False)

In [ ]:
importance

Feature  Coefficient  Importance
2           CholCheck    -1.258583    1.258583
5              Stroke     1.127289    1.127289
11      AnyHealthcare    -0.786991    0.786991
1            HighChol     0.680833    0.680833
17                Sex     0.675306    0.675306
10  HvyAlcoholConsump    -0.600276    0.600276
0              HighBP     0.549668    0.549668
4              Smoker     0.454665    0.454665
13            GenHlth     0.440967    0.440967
18                Age     0.228299    0.228299
8              Fruits    -0.209989    0.209989
6            Diabetes     0.184392    0.184392
16           DiffWalk     0.176273    0.176273
12        NoDocbcCost    -0.163976    0.163976
20             Income    -0.063639    0.063639
9             Veggies    -0.061593    0.061593
19          Education    -0.057693    0.057693
7        PhysActivity     0.028165    0.028165
3                 BMI    -0.017824    0.017824
15           PhysHlth     0.004466    0.004466
14           MentHlth     0.001101    0.001101

In [ ]:
a_dict = {}
for p in all_columns:
  X = heart_data[[p]]
  y = heart_data[target]
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
  lg = LogisticRegression()
  lg.fit(X_train, y_train)
  y_pred = lg.predict(X_test)
  accuracy = accuracy_score(y_test, y_pred)
  f1 = f1_score(y_test, y_pred)
  a_dict[p] = accuracy

a_dict = sorted(a_dict.items(), key= lambda x: x[1])
print(f'{a_dict[-1][0]} had the highest score of {a_dict[-1][1]}')

Income had the highest score of 0.9060233364869127


In [ ]:
sorted(a_dict, key=lambda x: x[1], reverse=True)

[('HighBP', 0.9060233364869127),
 ('HighChol', 0.9060233364869127),
 ('CholCheck', 0.9060233364869127),
 ('BMI', 0.9060233364869127),
 ('Smoker', 0.9060233364869127),
 ('Stroke', 0.9060233364869127),
 ('Diabetes', 0.9060233364869127),
 ('PhysActivity', 0.9060233364869127),
 ('Fruits', 0.9060233364869127),
 ('Veggies', 0.9060233364869127),
 ('HvyAlcoholConsump', 0.9060233364869127),
 ('AnyHealthcare', 0.9060233364869127),
 ('NoDocbcCost', 0.9060233364869127),
 ('GenHlth', 0.9060233364869127),
 ('MentHlth', 0.9060233364869127),
 ('PhysHlth', 0.9060233364869127),
 ('DiffWalk', 0.9060233364869127),
 ('Sex', 0.9060233364869127),
 ('Age', 0.9060233364869127),
 ('Education', 0.9060233364869127),
 ('Income', 0.9060233364869127)]

In [ ]:
#Testing a combination of the most important features ranked by the coefficients
X = heart_data[['Stroke', 'CholCheck', 'AnyHealthcare', 'HighChol', 'Sex', 'HvyAlcoholConsump', 'HighBP', 'Smoker', 'GenHlth']]
y = heart_data[target]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

lg = LogisticRegression()
lg.fit(X_train, y_train)
y_pred = lg.predict(X_test)

In [ ]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.91      0.99      0.95     45968
         1.0       0.54      0.11      0.18      4768

    accuracy                           0.91     50736
   macro avg       0.73      0.55      0.57     50736
weighted avg       0.88      0.91      0.88     50736



In [ ]:
print(heart_data[target].value_counts())

HeartDiseaseorAttack
0.0    229787
1.0     23893
Name: count, dtype: int64


#Neural Network

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, f1_score
from sklearn.preprocessing import StandardScaler

In [ ]:
all_columns = heart_data.columns.drop(target)

In [ ]:
X = heart_data[all_columns].values
y = heart_data[target].values

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
class NeuralNetwork:
    def __init__(self, num_feats):
        np.random.seed(1)
        self.synaptic_weights = 2 * np.random.random((num_feats, 1)) - 1

    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def sigmoid_derivative(self, x):
        return x * (1 - x)

    def train(self, training_inputs, training_outputs, training_iterations):
        for iteration in range(training_iterations):
            output = self.think(training_inputs)
            error = training_outputs - output
            adjustments = np.dot(training_inputs.T, error * self.sigmoid_derivative(output))
            self.synaptic_weights += adjustments

    def think(self, inputs):
        inputs = inputs.astype(float)
        output = self.sigmoid(np.dot(inputs, self.synaptic_weights))
        return output

In [ ]:
neural_network = NeuralNetwork(num_feats=len(all_columns))

print(f"Random starting synaptic weights: \n{neural_network.synaptic_weights}")

training_inputs = X_train
training_outputs = y_train.reshape(-1, 1)

neural_network.train(training_inputs, training_outputs, 1000)

print("Synaptic weights after training: ")
print(neural_network.synaptic_weights)

Random starting synaptic weights: 
[[-0.16595599]
 [ 0.44064899]
 [-0.99977125]
 [-0.39533485]
 [-0.70648822]
 [-0.81532281]
 [-0.62747958]
 [-0.30887855]
 [-0.20646505]
 [ 0.07763347]
 [-0.16161097]
 [ 0.370439  ]
 [-0.5910955 ]
 [ 0.75623487]
 [-0.94522481]
 [ 0.34093502]
 [-0.1653904 ]
 [ 0.11737966]
 [-0.71922612]
 [-0.60379702]
 [ 0.60148914]]


<ipython-input-32-d106be609126>:7: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-x))


In [ ]:
test_inputs = X_test
test_outputs = y_test.reshape(-1, 1)

raw_test_pred = neural_network.think(test_inputs)
test_pred = np.round(raw_test_pred).astype(int)
accuracy = accuracy_score(test_outputs, test_pred)

print(f'Accuracy: {accuracy_score(test_outputs, test_pred)}')
print(classification_report(y_test, test_pred))

<ipython-input-32-d106be609126>:7: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-x))


Accuracy: 0.633100756859035
              precision    recall  f1-score   support

         0.0       0.97      0.62      0.75     45968
         1.0       0.18      0.81      0.29      4768

    accuracy                           0.63     50736
   macro avg       0.57      0.71      0.52     50736
weighted avg       0.89      0.63      0.71     50736



#Tensorflow

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [ ]:
heart_data.columns

Index(['HeartDiseaseorAttack', 'HighBP', 'HighChol', 'CholCheck', 'BMI',
       'Smoker', 'Stroke', 'Diabetes', 'PhysActivity', 'Fruits', 'Veggies',
       'HvyAlcoholConsump', 'AnyHealthcare', 'NoDocbcCost', 'GenHlth',
       'MentHlth', 'PhysHlth', 'DiffWalk', 'Sex', 'Age', 'Education',
       'Income'],
      dtype='object')

In [ ]:
tf.random.set_seed(42)

In [ ]:
def remov(*items):
    return [item for item in list(heart_data.columns.drop(target)) if item not in items]

In [ ]:
all_columns = remov()

In [ ]:
X = heart_data[all_columns]
y = heart_data[target]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
ts = tf.keras.Sequential([
    tf.keras.layers.Dense(64, activation='relu', input_shape=(len(all_columns),)),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

ts.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001), loss='binary_crossentropy', metrics=['accuracy'])
ts.fit(X_train, y_train, epochs=10, batch_size=10)

Epoch 1/10
20295/20295 [==============================] - 49s 2ms/step - loss: 0.2527 - accuracy: 0.9057
Epoch 2/10
20295/20295 [==============================] - 48s 2ms/step - loss: 0.2439 - accuracy: 0.9070
Epoch 3/10
20295/20295 [==============================] - 49s 2ms/step - loss: 0.2418 - accuracy: 0.9073
Epoch 4/10
20295/20295 [==============================] - 47s 2ms/step - loss: 0.2408 - accuracy: 0.9073
Epoch 5/10
20295/20295 [==============================] - 48s 2ms/step - loss: 0.2403 - accuracy: 0.9075
Epoch 6/10
20295/20295 [==============================] - 46s 2ms/step - loss: 0.2400 - accuracy: 0.9077
Epoch 7/10
20295/20295 [==============================] - 47s 2ms/step - loss: 0.2394 - accuracy: 0.9076
Epoch 8/10
20295/20295 [==============================] - 49s 2ms/step - loss: 0.2392 - accuracy: 0.9077
Epoch 9/10
20295/20295 [==============================] - 46s 2ms/step - loss: 0.2391 - accuracy: 0.9077
Epoch 10/10
20295/20295 [==============================

In [ ]:
loss, accuracy = ts.evaluate(X_test, y_test)
print(f"Accuracy: {accuracy*100:.2f}%")

1586/1586 [==============================] - 3s 2ms/step - loss: 0.2376 - accuracy: 0.9082
Accuracy: 90.82%


In [ ]:
from sklearn.metrics import classification_report

y_pred = (ts.predict(X_test) > 0.5).astype(int)
print(classification_report(y_test, y_pred))

1586/1586 [==============================] - 3s 2ms/step
              precision    recall  f1-score   support

         0.0       0.92      0.99      0.95     45968
         1.0       0.55      0.12      0.19      4768

    accuracy                           0.91     50736
   macro avg       0.73      0.55      0.57     50736
weighted avg       0.88      0.91      0.88     50736



With all features: 82%

Without the few least important features evaluated from logistic model: 83%

#Other Models

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, f1_score, accuracy_score, mean_squared_error, r2_score, recall_score, precision_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV

In [ ]:
heart_data.columns.drop(target)

Index(['HighBP', 'HighChol', 'CholCheck', 'BMI', 'Smoker', 'Stroke',
       'Diabetes', 'PhysActivity', 'Fruits', 'Veggies', 'HvyAlcoholConsump',
       'AnyHealthcare', 'NoDocbcCost', 'GenHlth', 'MentHlth', 'PhysHlth',
       'DiffWalk', 'Sex', 'Age', 'Education', 'Income'],
      dtype='object')

In [ ]:
all_columns = list(heart_data.columns.drop(target))
X = heart_data[all_columns]
y = heart_data[target]

In [ ]:
#initializing 5 models
rf = RandomForestClassifier(max_depth=4,random_state=42, class_weight='balanced')
et = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
dt = DecisionTreeClassifier(max_depth=5, random_state=42, class_weight='balanced')
knn = KNeighborsClassifier(n_neighbors=5)
svm = SVC(kernel='linear', C=1, random_state=42)
gbm = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)

In [ ]:
#training and testing results lists to store in
accuracy = []
f1 = []
precision = []
recall = []
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
#function for training and testing
def training(model, aX_train, aX_test, ay_train, ay_test):
    model.fit(aX_train, ay_train)
    ay_train_pred = model.predict(aX_train)
    ay_test_pred = model.predict(aX_test)
    accuracy.append(accuracy_score(ay_test, ay_test_pred))
    f1.append(f1_score(ay_test, ay_test_pred))
    precision.append(precision_score(ay_test, ay_test_pred))
    recall.append(recall_score(ay_test, ay_test_pred))
    print(classification_report(ay_test, ay_test_pred))

In [ ]:
#calling training function for each model
training(rf, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

         0.0       0.97      0.71      0.82     45968
         1.0       0.23      0.81      0.35      4768

    accuracy                           0.72     50736
   macro avg       0.60      0.76      0.59     50736
weighted avg       0.90      0.72      0.78     50736



In [ ]:
training(dt, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

         0.0       0.97      0.71      0.82     45968
         1.0       0.22      0.79      0.34      4768

    accuracy                           0.72     50736
   macro avg       0.60      0.75      0.58     50736
weighted avg       0.90      0.72      0.78     50736



In [ ]:
training(knn, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

         0.0       0.91      0.98      0.95     45968
         1.0       0.35      0.11      0.17      4768

    accuracy                           0.90     50736
   macro avg       0.63      0.54      0.56     50736
weighted avg       0.86      0.90      0.87     50736



In [ ]:
training(gbm, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

         0.0       0.92      0.99      0.95     45968
         1.0       0.56      0.11      0.19      4768

    accuracy                           0.91     50736
   macro avg       0.74      0.55      0.57     50736
weighted avg       0.88      0.91      0.88     50736



In [ ]:
training(svm, aX_train, aX_test, ay_train, ay_test)

In [ ]:
training(et, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

         0.0       0.92      0.97      0.94     45968
         1.0       0.34      0.14      0.20      4768

    accuracy                           0.89     50736
   macro avg       0.63      0.55      0.57     50736
weighted avg       0.86      0.89      0.87     50736



In [ ]:
#all results
models = ["Random Forest", "Decision Tree", "k-Nearest Neighbors", "Gradient Boosting Machine", "Extra Trees"]
results = pd.DataFrame({
    "Model": models,
    "Accuracy": accuracy,
    "Precision": precision,
    "Recall": recall,
    "F1": f1
})
print(results)

                       Model  Accuracy  Precision    Recall        F1
0              Random Forest  0.722091   0.226109  0.807886  0.353330
1              Decision Tree  0.717715   0.220152  0.788171  0.344171
2        k-Nearest Neighbors  0.897134   0.348352  0.108641  0.165627
3  Gradient Boosting Machine  0.908487   0.564633  0.114513  0.190410
4                Extra Trees  0.894119   0.341719  0.136745  0.195327


In [ ]:
heart_data.columns

Index(['HeartDiseaseorAttack', 'HighBP', 'HighChol', 'CholCheck', 'BMI',
       'Smoker', 'Stroke', 'Diabetes', 'PhysActivity', 'Fruits', 'Veggies',
       'HvyAlcoholConsump', 'AnyHealthcare', 'NoDocbcCost', 'GenHlth',
       'MentHlth', 'PhysHlth', 'DiffWalk', 'Sex', 'Age', 'Education',
       'Income'],
      dtype='object')

##Feature Importance Ranking

In [ ]:
def rank(the_model):
  feature_importances = the_model.feature_importances_
  ranking = pd.DataFrame({
      'Feature': X.columns,
      'Importance': feature_importances
  }).sort_values(by='Importance', ascending=False)

  print(ranking)

In [ ]:
rank(rf)

              Feature    Importance
13            GenHlth  2.184990e-01
18                Age  1.830655e-01
0              HighBP  1.817854e-01
1            HighChol  1.054481e-01
16           DiffWalk  8.269268e-02
6            Diabetes  5.820711e-02
15           PhysHlth  5.051277e-02
5              Stroke  4.652392e-02
20             Income  2.433568e-02
4              Smoker  2.389483e-02
17                Sex  1.579854e-02
19          Education  5.608967e-03
14           MentHlth  9.187609e-04
7        PhysActivity  8.526712e-04
3                 BMI  7.545488e-04
2           CholCheck  5.879447e-04
10  HvyAlcoholConsump  2.407890e-04
11      AnyHealthcare  2.097566e-04
12        NoDocbcCost  4.427300e-05
9             Veggies  1.819106e-05
8              Fruits  6.341855e-07


In [ ]:
rank(dt)

              Feature  Importance
0              HighBP    0.399613
13            GenHlth    0.240273
18                Age    0.233080
17                Sex    0.050645
1            HighChol    0.036845
5              Stroke    0.029900
4              Smoker    0.007804
16           DiffWalk    0.001505
15           PhysHlth    0.000334
12        NoDocbcCost    0.000000
19          Education    0.000000
14           MentHlth    0.000000
10  HvyAlcoholConsump    0.000000
11      AnyHealthcare    0.000000
9             Veggies    0.000000
8              Fruits    0.000000
7        PhysActivity    0.000000
6            Diabetes    0.000000
3                 BMI    0.000000
2           CholCheck    0.000000
20             Income    0.000000


In [ ]:
rank(gbm)

              Feature  Importance
13            GenHlth    0.305291
18                Age    0.206707
5              Stroke    0.103869
0              HighBP    0.095155
17                Sex    0.076869
16           DiffWalk    0.074185
1            HighChol    0.070609
6            Diabetes    0.028254
4              Smoker    0.019011
20             Income    0.005704
15           PhysHlth    0.005315
3                 BMI    0.002581
10  HvyAlcoholConsump    0.002124
12        NoDocbcCost    0.001563
14           MentHlth    0.001019
19          Education    0.000650
2           CholCheck    0.000647
9             Veggies    0.000243
7        PhysActivity    0.000203
11      AnyHealthcare    0.000000
8              Fruits    0.000000


In [ ]:
rank(et)

In [ ]:
from sklearn.inspection import permutation_importance

In [ ]:
knn_result = permutation_importance(knn, X_test, y_test, n_repeats=10, random_state=42)

In [ ]:
svm_result = permutation_importance(svm, X_test, y_test, n_repeats=10, random_state=42)

In [ ]:
def rank_2(result):
  perms = pd.DataFrame({
      'Feature': X.columns,
      '(Mean) Importance': result.importances_mean,
  })
  sorted_importances = perms.sort_values(by='(Mean) Importance', ascending=False)

  print(sorted_importances)

In [ ]:
rank_2(knn_result)

                Feature  (Mean) Importance
7        max heart rate           0.137815
4           cholesterol           0.101261
3          resting bp s           0.042437
0                   age           0.020588
8       exercise angina           0.003361
6           resting ecg           0.002941
9               oldpeak           0.002941
1                   sex           0.001681
10             ST slope           0.000000
5   fasting blood sugar          -0.000840
2       chest pain type          -0.002941


In [ ]:
rank_2(svm_result)

                Feature  (Mean) Importance
10             ST slope           0.079412
2       chest pain type           0.037815
8       exercise angina           0.035294
1                   sex           0.025210
9               oldpeak           0.015966
4           cholesterol           0.003361
5   fasting blood sugar           0.001261
6           resting ecg          -0.002521
3          resting bp s          -0.002941
7        max heart rate          -0.005042
0                   age          -0.013866


##Testing collective performance by removing certain features from all features

In [ ]:
initial_dict = {} #run this only once

In [ ]:
def remov(*items):
    return [item for item in list(heart_data.columns.drop(target)) if item not in items]

In [ ]:
all_columns = remov('exercise angina') #'slope' had a good effect on GBM

In [ ]:
X = heart_data[all_columns]
y = heart_data[target]
gbm = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)
#training and testing results lists to store in
accuracy = []
f1 = []
precision = []
recall = []
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
def training(model, aX_train, aX_test, ay_train, ay_test):
    model.fit(aX_train, ay_train)
    ay_train_pred = model.predict(aX_train)
    ay_test_pred = model.predict(aX_test)
    accuracy.append(accuracy_score(ay_test, ay_test_pred))
    f1.append(f1_score(ay_test, ay_test_pred))
    precision.append(precision_score(ay_test, ay_test_pred))
    recall.append(recall_score(ay_test, ay_test_pred))
training(rf, aX_train, aX_test, ay_train, ay_test)
training(dt, aX_train, aX_test, ay_train, ay_test)
training(knn, aX_train, aX_test, ay_train, ay_test)
training(gbm, aX_train, aX_test, ay_train, ay_test)
training(svm, aX_train, aX_test, ay_train, ay_test)
training(et, aX_train, aX_test, ay_train, ay_test)
#one model's results over different features
models = ["Random Forest", "Decision Tree", "k-Nearest Neighbors", "Gradient Boosting Machine", "Support Vector Machine", "Extra Trees"]
results = pd.DataFrame({
    "Model": models,
    "Accuracy": accuracy,
    "Precision": precision,
    "Recall": recall,
    "F1": f1
})
print(results)
initial_dict[str(all_columns)+" | length "+str(len(all_columns))] = results

                       Model  Accuracy  Precision    Recall        F1
0              Random Forest  0.886555   0.893939  0.900763  0.897338
1              Decision Tree  0.848739   0.818792  0.931298  0.871429
2        k-Nearest Neighbors  0.718487   0.753968  0.725191  0.739300
3  Gradient Boosting Machine  0.907563   0.903704  0.931298  0.917293
4     Support Vector Machine  0.831933   0.837037  0.862595  0.849624
5                Extra Trees  0.936975   0.946154  0.938931  0.942529


In [ ]:
initial_dict

{"['age', 'sex', 'chest pain type', 'resting bp s', 'cholesterol', 'fasting blood sugar', 'resting ecg', 'max heart rate', 'exercise angina', 'oldpeak', 'ST slope']":                        Model  Accuracy  Precision    Recall        F1
 0              Random Forest  0.882353   0.893130  0.893130  0.893130
 1              Decision Tree  0.848739   0.818792  0.931298  0.871429
 2        k-Nearest Neighbors  0.718487   0.753968  0.725191  0.739300
 3  Gradient Boosting Machine  0.915966   0.917293  0.931298  0.924242
 4     Support Vector Machine  0.848739   0.857143  0.870229  0.863636
 5                Extra Trees  0.936975   0.953125  0.931298  0.942085,
 "['age', 'chest pain type', 'resting bp s', 'cholesterol', 'resting ecg', 'max heart rate', 'exercise angina', 'oldpeak', 'ST slope']":                        Model  Accuracy  Precision    Recall        F1
 0              Random Forest  0.865546   0.866667  0.893130  0.879699
 1              Decision Tree  0.815126   0.791946  0.9007

Looks like removing exercise angina increased the accuracy by a tiny bit

##Testing for best single feature for the best models



In [ ]:
#initializing 5 models
rf = RandomForestClassifier(max_depth=4,random_state=42, class_weight='balanced')
et = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
dt = DecisionTreeClassifier(max_depth=5, random_state=42, class_weight='balanced')
knn = KNeighborsClassifier(n_neighbors=5)
svm = SVC(kernel='linear', C=1, random_state=42)
gbm = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)

In [ ]:
def single_feature(c_model):
  for k in heart_data.columns.drop(target):
    X = heart_data[[k]]
    y = heart_data[target]
    #f1 score
    f1 = 0
    #splitting data
    aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
    #function for training and testing
    c_model.fit(aX_train, ay_train)
    ay_train_pred = c_model.predict(aX_train)
    ay_test_pred = c_model.predict(aX_test)
    f1 = f1_score(ay_test, ay_test_pred)
    #one model's results over different features
    print(k+":", f1)

In [ ]:
single_feature(rf)

age: 0.6381322957198444
sex: 0.7562500000000001
chest pain type: 0.7795275590551182
resting bp s: 0.4519230769230769
cholesterol: 0.6180257510729613
fasting blood sugar: 0.42222222222222217
resting ecg: 0.5407725321888412
max heart rate: 0.723076923076923
exercise angina: 0.7381974248927038
oldpeak: 0.7086614173228347
ST slope: 0.8074074074074076


In [ ]:
single_feature(gbm)

age: 0.640926640926641
sex: 0.7562500000000001
chest pain type: 0.7795275590551182
resting bp s: 0.6343283582089552
cholesterol: 0.688
fasting blood sugar: 0.42222222222222217
resting ecg: 0.5407725321888412
max heart rate: 0.7272727272727273
exercise angina: 0.7381974248927038
oldpeak: 0.7086614173228347
ST slope: 0.8074074074074076


In [ ]:
single_feature(et)

age: 0.6381322957198444
sex: 0.7562500000000001
chest pain type: 0.7795275590551182
resting bp s: 0.5405405405405406
cholesterol: 0.6422764227642276
fasting blood sugar: 0.42222222222222217
resting ecg: 0.5407725321888412
max heart rate: 0.6746031746031745
exercise angina: 0.7381974248927038
oldpeak: 0.7086614173228347
ST slope: 0.8074074074074076


##Concluding Evaluations

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
heart_data.columns.drop(target)

Index(['age', 'sex', 'chest pain type', 'resting bp s', 'cholesterol',
       'fasting blood sugar', 'resting ecg', 'max heart rate',
       'exercise angina', 'oldpeak', 'ST slope'],
      dtype='object')

Lifestyle factor(s): exercise angina

###With lifestyle factor(s)

In [ ]:
#testing for the best mulitple features ranked by importance by that model, from a preivous section
all_columns = remov('exercise angina')
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
with_f1 = f1_score(ay_test,ay_test_pred)

###Without lifestyle factor(s)

In [ ]:
#Using best model from preivous evaluations
all_columns = remov()
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
without_f1 = f1_score(ay_test,ay_test_pred)

In [ ]:
print(f'{with_f1} (factors)\n{without_f1} (no factors)')

0.9425287356321839 (factors)
0.9420849420849421 (no factors)


###Findings (Lifestyle vs No Lifestyle Factors)
Removing exercise induced angina did not affect the result much, as it barely decreased the accuracy.

However, the feature importance of Extra Trees shows that it was the second most important feature, second to ST Slope.

Removing Other Most Important Features Ranked from Before for this model:

ST slope (1st), chest pain (3rd), old peak (4th)

In [ ]:
#Using best model from preivous evaluations
all_columns = remov('ST slope')
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
print(f1_score(ay_test,ay_test_pred))

0.9140625000000001


In [ ]:
#Using best model from preivous evaluations
all_columns = remov('chest pain')
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
print(f1_score(ay_test,ay_test_pred))

0.9420849420849421


In [ ]:
#Using best model from preivous evaluations
all_columns = remov('oldpeak')
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
print(f1_score(ay_test,ay_test_pred))

0.9501915708812262


* Removing oldpeak improved the model performance by 1%
* Removing chest pain did not have much of an impact
* Removing ST slope decreased accuracy to 91%

Therefore, ST slope seems like an an important feature in predicting heart disease, as it was also ranked the most important feature of this model

Slope of the Peak Exercise ST Segment is a psycological measure that can indicate how well the heart is functioning under stress.